File: corpus/apps/kmp/src/commonMain/kotlin/com/orbitalfreight/shared/ContainerRegistryClient.kt
package com.orbitalfreight.shared

import com.orbitalfreight.shared.model.Container
import com.orbitalfreight.shared.model.IdempotencyKeys
import com.orbitalfreight.shared.model.Position
import com.orbitalfreight.shared.model.Scan
import com.orbitalfreight.shared.model.Shipment
import io.ktor.client.request.parameter
import io.ktor.http.HttpMethod
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * container-registry (§3.3, HTTP 8083) का क्लाइंट — शिपमेंट, डिब्बे और स्कैन-पथ।
 *
 * यह उस सेवा का *पूरा* मुख वाला हिस्सा नहीं है, केवल वह जो मोबाइल से चलता है।
 * तीन पथ जान-बूझकर बाहर हैं:
 *
 *  • `POST /v1/containers` — डिब्बा पंजीकरण डिपो कार्यालय का काम है।
 *  • `DELETE /v1/shipments/{shipment_id}/containers/{container_id}` — सील लगने
 *    से पहले ही चलता है और केवल कंसोल से।
 *  • `freight.v1.ContainerLookup/ResolveShipmentForContainer` — वह gRPC है और
 *    telemetry-ingest के गरम रास्ते के लिए है; मोबाइल उसे कभी नहीं छूता।
 *
 * स्थिति बदलने का एक ही वैध रास्ता `PATCH …/status` है। ऐप कहीं और से स्थिति
 * नहीं बदल सकता, और यह सीमा सर्वर की है — यहाँ केवल दोहराई गई है ताकि पढ़ने
 * वाला दूसरा रास्ता ढूँढने न निकले।
 */
class ContainerRegistryClient(private val http: OrbitalHttpClient) {

    /** `GET /v1/shipments/{shipment_id}` — डिब्बों सहित पूरा शिपमेंट। */
    suspend fun shipment(shipmentId: String, traceId: String? = null): Shipment =
        http.send(HttpMethod.Get, "/v1/shipments/$shipmentId", traceId = traceId)

    /**
     * `GET /v1/containers` — BIC कोड, सील या शिपमेंट से खोज।
     *
     * तीनों फ़िल्टर एक साथ देने का कोई अर्थ नहीं है और सर्वर उन्हें AND की तरह
     * लेता है; व्यवहार में स्कैनर हमेशा एक ही देता है — वही जो बारकोड में था।
     */
    suspend fun findContainers(
        isoCode: String? = null,
        seal: String? = null,
        shipmentId: String? = null,
        limit: Int = OrbitalHttpClient.DEFAULT_PAGE_LIMIT,
        cursor: String? = null,
    ): Page<Container> {
        require(limit in 1..OrbitalHttpClient.MAX_PAGE_LIMIT) {
            "limit must be between 1 and ${OrbitalHttpClient.MAX_PAGE_LIMIT} (§0.5)"
        }
        return http.send(HttpMethod.Get, "/v1/containers") {
            isoCode?.let { parameter("iso_code", it) }
            seal?.let { parameter("seal", it) }
            shipmentId?.let { parameter("shipment_id", it) }
            parameter("limit", limit)
            cursor?.let { parameter("cursor", it) }
        }
    }

    /**
     * `POST /v1/containers/{container_id}/scans` — एक स्कैन दर्ज करता है।
     *
     * सफल होने पर सेवा `shipment.scanned` को `platform.outbox_messages` में
     * लिखती है और उसका relay उसे `of.freight.v1` पर प्रकाशित करता है। ऐप उस
     * घटना का उपभोक्ता नहीं है — जो कुछ उसे वापस चाहिए वह इसी उत्तर में है।
     *
     * @param idempotencyKey कतार की पंक्ति के साथ बँधी कुंजी। हर पुनःप्रयास पर
     *        वही भेजनी है; नई कुंजी का अर्थ है वही स्कैन दूसरी बार दर्ज होना।
     */
    suspend fun recordScan(
        containerId: String,
        request: RecordScanRequest,
        idempotencyKey: String,
        traceId: String? = null,
    ): Scan = http.send(
        method = HttpMethod.Post,
        path = "/v1/containers/$containerId/scans",
        body = request,
        idempotencyKey = idempotencyKey,
        traceId = traceId,
    )

    /** `GET /v1/shipments/{shipment_id}/scans` — स्कैन का इतिहास, नया पहले। */
    suspend fun scanTrail(
        shipmentId: String,
        limit: Int = OrbitalHttpClient.DEFAULT_PAGE_LIMIT,
        cursor: String? = null,
    ): Page<Scan> = http.send(HttpMethod.Get, "/v1/shipments/$shipmentId/scans") {
        parameter("limit", limit)
        cursor?.let { parameter("cursor", it) }
    }

    /**
     * `PATCH /v1/shipments/{shipment_id}/status` — स्थिति बदलने का एकमात्र रास्ता।
     *
     * मोबाइल से केवल दो संक्रमण चलते हैं और दोनों गोदाम/गेट पर होते हैं:
     * `booked` → `sealed` और `sealed` → `in_transit`। बाकी या तो
     * `customs.declaration.*` घटनाओं से आते हैं, या चेतावनी से (`at_risk`), या
     * कंसोल से। सर्वर वैसे भी अवैध संक्रमण को `409` देता है — यह टिप्पणी इसलिए
     * है कि कोई नई स्क्रीन बनाते समय तीसरा संक्रमण जोड़ने से पहले रुक जाए।
     */
    suspend fun changeStatus(
        shipmentId: String,
        toStatus: String,
        reasonCode: String,
        idempotencyKey: String = IdempotencyKeys.newIdempotencyKey(),
        traceId: String? = null,
    ): Shipment = http.send(
        method = HttpMethod.Patch,
        path = "/v1/shipments/$shipmentId/status",
        body = ChangeStatusRequest(status = toStatus, reasonCode = reasonCode),
        idempotencyKey = idempotencyKey,
        traceId = traceId,
    )
}

/**
 * स्कैन का शरीर।
 *
 * `recorded_at` यहाँ नहीं है और होना भी नहीं चाहिए — वह सर्वर की घड़ी है।
 * `scanned_by_user_id` भी नहीं: वह टोकन से आता है, और उसे शरीर में भेजने देना
 * एक ऐसा रास्ता खोलता है जिससे कोई दूसरे के नाम स्कैन दर्ज कर सके।
 */
@Serializable
data class RecordScanRequest(
    @SerialName("shipment_id") val shipmentId: String,
    @SerialName("scan_type") val scanType: String,
    @SerialName("occurred_at") val occurredAt: String,
    @SerialName("facility_id") val facilityId: String? = null,
    val position: Position? = null,
    @SerialName("device_serial") val deviceSerial: String,
    val notes: String? = null,
)

@Serializable
data class ChangeStatusRequest(
    val status: String,
    /** यही `shipment.status.changed` में `reason_code` बनकर जाता है। */
    @SerialName("reason_code") val reasonCode: String,
)

File: corpus/services/portal/app/clients/container_registry_client.rb
# frozen_string_literal: true
#
# Client HTTP vers container-registry, système de référence des expéditions, des conteneurs et de
# la piste de scans (§1). Le portail lit ici tout ce qu'il affiche sur l'écran d'expédition et
# n'écrit que par l'unique chemin de transition d'état autorisé, PATCH /v1/shipments/{shipment_id}/status.

require_relative 'base_client'
require_relative '../../lib/portal/prefixed_ulid'

module Portal
  module Clients
    class ContainerRegistryClient < BaseClient
      def self.service_name
        'container-registry'
      end

      # Transitions que l'écran propose réellement. container-registry reste seul juge — cette
      # table n'est qu'un filtre d'ergonomie, elle évite de présenter un bouton qui produira un
      # 409 `illegal_status_transition`. Les statuts viennent du CHECK sur freight.shipments.
      ALLOWED_TRANSITIONS = {
        'draft'           => %w[booked cancelled],
        'booked'          => %w[sealed cancelled],
        'sealed'          => %w[in_transit cancelled],
        'in_transit'      => %w[at_risk held_at_customs delivered],
        'at_risk'         => %w[in_transit held_at_customs delivered],
        'held_at_customs' => %w[in_transit cancelled],
        'delivered'       => [],
        'cancelled'       => []
      }.freeze

      # `GET /v1/shipments/{shipment_id}` renvoie déjà les conteneurs en ligne ; inutile de faire
      # un second appel sur /containers pour peupler le tableau de l'écran.
      def shipment(shipment_id)
        PrefixedUlid.assert!(shipment_id, :shipment)
        get("/v1/shipments/#{shipment_id}")
      end

      def scan_trail(shipment_id, limit: 100)
        PrefixedUlid.assert!(shipment_id, :shipment)
        # La route rend déjà le plus récent en premier ; on ne retrie pas côté portail, sans quoi
        # deux scans partageant la même seconde changeraient d'ordre à chaque rafraîchissement.
        paginate("/v1/shipments/#{shipment_id}/scans", {}, max_items: limit)
      end

      def container(container_id)
        PrefixedUlid.assert!(container_id, :container)
        get("/v1/containers/#{container_id}")
      end

      # `GET /v1/containers` filtre par iso_code, seal ou shipment_id — et rien d'autre. Une
      # recherche par numéro de scellé passe donc par `seal`, pas par une recherche plein texte
      # qui n'existe nulle part dans la plateforme.
      def containers_by_seal(seal_number)
        get('/v1/containers', seal: seal_number).fetch('items', [])
      end

      # container-registry n'expose aucune route de liste sur /v1/shipments : §3.3 ne connaît que
      # la lecture unitaire GET /v1/shipments/{shipment_id}. C'est délibéré — le système de
      # référence ne veut pas d'un balayage plein tenant. Les listes du portail partent donc
      # toujours d'un objet qui porte déjà shipment_id : une alerte de telemetry-ingest, une
      # facture de billing-service, un écart de reconciliation-service.
      def shipments_by_ids(shipment_ids)
        shipment_ids.uniq.take(25).filter_map do |id|
          shipment(id)
        rescue UpstreamError => e
          # Une expédition purgée ou appartenant à un autre tenant ne doit pas faire tomber la
          # liste entière : on la saute et on continue.
          Portal.logger.debug(msg: 'expédition ignorée dans la liste', shipment_id: id, code: e.code)
          nil
        end
      end

      # Conteneurs rattachés à une expédition. §3.3 : GET /v1/containers filtre sur iso_code,
      # seal et shipment_id, et rien d'autre.
      def containers_for_shipment(shipment_id)
        PrefixedUlid.assert!(shipment_id, :shipment)
        get('/v1/containers', shipment_id: shipment_id).fetch('items', [])
      end

      def containers_by_iso_code(iso_code)
        get('/v1/containers', iso_code: iso_code.to_s.upcase).fetch('items', [])
      end

      # Seul chemin d'écriture d'état. La clé d'idempotence est dérivée du couple
      # (expédition, statut visé) plutôt que tirée au hasard : deux opérateurs qui cliquent
      # simultanément « passer en transit » doivent produire une transition, pas deux entrées
      # dans platform.audit_ledger_entries.
      def change_status(shipment_id, to_status:, reason_code:)
        PrefixedUlid.assert!(shipment_id, :shipment)

        patch(
          "/v1/shipments/#{shipment_id}/status",
          { status: to_status, reason_code: reason_code },
          idempotency_key: "portal-status-#{shipment_id}-#{to_status}"
        )
      end

      # Détacher un conteneur n'est possible qu'avant le scellement — au-delà, container-registry
      # répond 409 `shipment_already_sealed`, l'exemple même de §0.4.
      def detach_container(shipment_id, container_id)
        PrefixedUlid.assert!(shipment_id, :shipment)
        PrefixedUlid.assert!(container_id, :container)

        request_delete("/v1/shipments/#{shipment_id}/containers/#{container_id}")
      end

      # Vue consolidée d'un écran d'expédition : la fiche, la piste de scans, et le drapeau
      # « livrée » que billing-service attend pour facturer. Cette dernière information ne sort
      # pas d'un champ dédié : c'est la présence d'un scan `proof_of_delivery` qui fait foi,
      # exactement comme dans le consommateur de `shipment.scanned` côté billing-service (§4.4).
      def shipment_overview(shipment_id)
        record = shipment(shipment_id)
        trail = scan_trail(shipment_id, limit: 50)

        {
          'shipment' => record,
          'scans' => trail,
          'delivered_scan' => trail.find { |scan| scan['scan_type'] == 'proof_of_delivery' },
          'available_transitions' => ALLOWED_TRANSITIONS.fetch(record['status'], [])
        }
      end

      private

      def request_delete(path)
        send(:request, Net::HTTP::Delete, path)
      end
    end
  end
end
